src/gs-background-tile.c

```diff
@@ -0,0 +1,281 @@
+	}
+
+	gradients_str = g_strjoinv (",", gradients);
+	css = g_strconcat ("background: ", gradients_str, ";", NULL);
+
+	gs_utils_widget_set_css (GTK_WIDGET (tile->image_box), css);
+	tile->key_colors_set = TRUE;
+}
+
+static void
+gs_background_tile_update_price (GsBackgroundTile *tile)
+{
+	const GsPrice *price = NULL;
+	g_autofree gchar *price_str = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		gtk_widget_hide (tile->price);
+		gtk_widget_show (tile->installed_icon);
+		return;
+	}
+
+	price = gs_app_get_price (tile->app);
+	if (price == NULL || (gs_price_get_amount (price) == (gdouble) 0)) {
+		/* TRANSLATORS: This is the price of the app, as in gratis */
+		price_str = g_utf8_strup (_("Free"), -1);
+		gtk_label_set_label (GTK_LABEL (tile->price), price_str);
+	} else {
+		price_str = gs_price_to_string (price);
+		gtk_label_set_label (GTK_LABEL (tile->price), price_str);
+	}
+
+	gtk_widget_show (tile->price);
+	gtk_widget_hide (tile->installed_icon);
+}
+
+static void
+update_tile_info (GsBackgroundTile *tile)
+{
+	g_autofree gchar *name = NULL;
+	AtkObject *accessible = gtk_widget_get_accessible (GTK_WIDGET (tile));
+	const gchar *css = NULL;
+	const gchar *summary = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		/* TRANSLATORS: This is the name and state of an app for the ATK object */
+		name = g_strdup_printf (_("%s (Installed)"), gs_app_get_name (tile->app));
+	} else if (gs_app_get_state (tile->app) == AS_APP_STATE_AVAILABLE) {
+		name = g_strdup (gs_app_get_name (tile->app));
+	}
+
+	summary = gs_app_get_summary (tile->app);
+
+	if (GTK_IS_ACCESSIBLE (accessible)) {
+		atk_object_set_name (accessible, name);
+		atk_object_set_description (accessible, summary);
+	}
+
+	gtk_label_set_label (GTK_LABEL (tile->name_label), gs_app_get_name (tile->app));
+	gtk_label_set_label (GTK_LABEL (tile->summary_label), summary);
+
+	css = gs_app_get_metadata_item (tile->app, "GnomeSoftware::BackgroundTile-css");
+	if (css != NULL)
+		gs_utils_widget_set_css (GTK_WIDGET (tile->image_box), css);
+	else
+		gs_background_tile_setup_colors_bg (tile);
+
+	gs_background_tile_update_price (tile);
+}
+
+static gboolean
+app_state_changed_idle (gpointer user_data)
+{
+	GsBackgroundTile *tile = GS_BACKGROUND_TILE (user_data);
+
+	update_tile_info (tile);
+	g_object_unref (tile);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+app_state_changed (GsApp *app, GParamSpec *pspec, GsBackgroundTile *tile)
+{
+	/* we call the function in idle because the state-change
+	 * notification could be coming from a different thread */
+	g_idle_add (app_state_changed_idle, g_object_ref (tile));
+}
+
+static void
+app_background_tile_css_changed (GsApp *app, const char *metadata, GsBackgroundTile *tile)
+{
+	/* just call the app_state_changed function as it will deal with the tile
+	 * CSS already; calling it from idle as this can come from a different thread */
+	g_idle_add (app_state_changed_idle, g_object_ref (tile));
```

Comment:
I don't believe metadata notification can happen in a different thread actually, or at least the previous code assumed that it did not.